pandas/core/generic.py

```diff
@@ -6308,11 +6308,23 @@ def fillna(
         else:
             if self.ndim == 1:
                 if isinstance(value, (dict, ABCSeries)):
-                    value = create_series_with_explicit_dtype(
+                    value_map = create_series_with_explicit_dtype(
                         value, dtype_if_empty=object
                     )
-                    value = value.reindex(self.index, copy=False)
-                    value = value._values
+                    # GH#40498 objects can have multiple types of missing values which
+                    # should not be modified unless specified.  Add special casing to
+                    # minimize performance decrease on other data types where this is
+                    # not required.
+                    if is_object_dtype(self.dtype):
+                        value = self.copy()
+                        modification_index = value.index.intersection(value_map.index)
+                        if not modification_index.empty:
+                            value.loc[modification_index] = value_map[
+                                modification_index
+                            ]
+                    else:
+                        value = value_map.reindex(self.index)
```

Comment:
if you don't add the copy does this just work e.g. do you still need L6314-6325

Reply:
This is currently not caught by the testing method:
```
result = pd.Series([1, None, 3, "four"])
expected = pd.Series([1, np.nan, 3, "four"])
tm.assert_equal(result, expected)
```

The unit tests will pass but the original problem will still exist.  I mentioned this on my initial Pull Request but these tests will not fail because of an open issue - #18463.  I probably should've clarified if it's worth writing tests with other bugs in mind, or to write them as if the bug will be fixed eventually.

Reply:
.reindex should just work here. I am really not sure what you are trying to do.

Reply:
can you update here